### PR TITLE
OA-42: Encode request parameter values in the logout url

### PR DESCRIPTION
**Pull request by wluyima**
[OA-42: Added property to disable encoding of post logout URL](https://openmrs.atlassian.net/browse/OA-42)

### CODE REVIEW

**Review by corneliouzbett — APPROVED** [review]
This LGTM! Let's add one more test to handle edge cases like empty query parameters or null parameters.

**Comment by wluyima** [comment]
Empty query parameter is the same as any other parameter value

File: omod/src/test/java/org/openmrs/module/oauth2login/web/UtilsTest.java [hunk 1]
```diff
@@ -71,4 +84,127 @@ public void onLogoutSuccess_shouldNotSetIdTokenIfNoneIsSetForTheUser() throws Ex
...
+		
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		
+		String redirect = Utils.getPostLogoutRedirectUrl(request);
+		
+		assertEquals("https://idp.com?" + PARAM_POST_LOGOUT_URL + "=http%3A%2F%2Fopenmrs.org", redirect);
+	}
+	
+	@Test
+	public void getPostLogoutRedirectUrl_shouldEncodeTheUrlIfDisabled() throws Exception {
```
Comment (wluyima) [inline omod/src/test/java/org/openmrs/module/oauth2login/web/UtilsTest.java:191]: Ooops! This should be **shouldNotEncodeUrlIfDisabled**

File: omod/src/main/java/org/openmrs/module/oauth2login/web/Utils.java [hunk 2]
```diff
@@ -56,7 +60,28 @@ public static String getPostLogoutRedirectUrl(HttpServletRequest request) throws
...
+		urlBuilderEncoded.host(urlComponents.getHost());
+		urlBuilderEncoded.port(urlComponents.getPort());
+		urlBuilderEncoded.path(urlComponents.getPath());
+		urlBuilderEncoded.fragment(urlComponents.getFragment());
+		if (params != null) {
+			for (Map.Entry<String, List<String>> entry : params.entrySet()) {
+				final String encodedKey = URLEncoder.encode(entry.getKey());
+				entry.getValue().forEach(v -> urlBuilderEncoded.queryParam(encodedKey, URLEncoder.encode(v)));
+			}
+		}
```
Comment (corneliouzbett) [inline omod/src/main/java/org/openmrs/module/oauth2login/web/Utils.java:83]: > Empty query parameter is effectively a blank parameter value which is the same as any other parameter value in my opinion.

I get your point, I was suggesting to capture (or unit test) the above functionality. What is the expected outcome if `params == null`, can we capture that in tests

Comment (wluyima) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/Utils.java:83]: I get your point, but is there an IDP that accepts a parameterless logout URL anyways? I intentionally omitted that test, in fact, for this reason I wanted to remove that if clause instead, params can never be null or empty otherwise that would be an invalid URL since the **client_id** or the **id_token_hint** parameter is required but this could be a Keycloak's implementation to require one of them, may be it is not the case for other implementations, so may be it make sense avoid a  potential NPE in case of other IDPs that might might accept a parameterless logout URL.

Comment (corneliouzbett) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/Utils.java:83]: ➕ it's a utility method, someone might want to use it to encode a URL that's not specific to Keycloak's implementation.
